packages/flutter_tools/test/integration/test_driver.dart

```diff
@@ -496,6 +506,85 @@ class FlutterRunTestDriver extends FlutterTestDriver {
   }
 }
 
+class FlutterTestTestDriver extends FlutterTestDriver {
+  FlutterTestTestDriver(Directory _projectFolder, {String logPrefix}):
+    super(_projectFolder, logPrefix: logPrefix);
+
+  Future<void> test({
+    String testFile = 'test/test.dart',
+    bool withDebugger = false,
+    bool pauseOnExceptions = false,
+    File pidFile,
+    Future<void> Function() beforeStart,
+  }) async {
+    await _setupProcess(<String>[
+        'test',
+        '--machine',
+        '-d',
+        'flutter-tester'
+    ], script: testFile, withDebugger: withDebugger, pauseOnExceptions: pauseOnExceptions, pidFile: pidFile);
+  }
+
+  @override
+  Future<void> _setupProcess(
+    List<String> args, {
+    String script,
+    bool withDebugger = false,
+    bool pauseOnExceptions = false,
+    File pidFile,
+    Future<void> Function() beforeStart,
+  }) async {
+    await super._setupProcess(
+      args,
+      script: script,
+      withDebugger: withDebugger,
+      pauseOnExceptions: pauseOnExceptions,
+      pidFile: pidFile,
+    );
+
+    // Stash the PID so that we can terminate the VM more reliably than using
+    // _proc.kill() (because _proc is a shell, because `flutter` is a shell
+    // script).
+    final Map<String, dynamic> version = await _waitForJson();
+    _procPid = version['pid'];
+
+    if (withDebugger) {
+      final Map<String, dynamic> startedProcess =
+          await _waitFor(event: 'test.startedProcess', timeout: appStartTimeout);
+      final String vmServiceHttpString = startedProcess['params']['observatoryUri'];
+      _vmServiceWsUri = Uri.parse(vmServiceHttpString).replace(scheme: 'ws', path: '/ws');
+      await connectToVmService(pauseOnExceptions: pauseOnExceptions);
+      // Allow us to run code before we start, eg. to set up breakpoints.
+      if (beforeStart != null) {
+        await beforeStart();
+      }
+      await resume(wait: false);
+    }
+  }
+
+  Future<Map<String, dynamic>> _waitForJson({
+    Duration timeout,
+  }) async {
+    return _timeoutWithMessages<Map<String, dynamic>>(
+      () => _stdout.stream.map<Map<String, dynamic>>(_parseJsonResponse).first,
+      timeout: timeout,
+      message: 'Did not receive any JSON.',
+    );
+  }
+
+  Map<String, dynamic> _parseJsonResponse(String line) {
+    if (line.startsWith('{') && line.endsWith('}')) {
```

Comment:
why not always try and parse the response?

Reply:
The output can contain non-JSON (eg. anything that's `print()`ed by the app) so it seemed a bit wasteful to catch exceptions for this (I think maybe I also considered that we may print in the exception case, but probably don't want to print all obvious non-JSON).

It doesn't seem like a solid reason now though, so I'll drop it out while tweaking the other bits :-)

Reply:
Actually, this one is `flutter test`, so maybe it's less likely than in Flutter run. Removed anyway!